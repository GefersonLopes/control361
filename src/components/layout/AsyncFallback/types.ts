export interface AsyncFallbackProps {
  isLoading: boolean;
  isError: boolean;
  errorContent?: React.ReactNode;
  loadingContent?: React.ReactNode;
  children: React.ReactNode;
  className?: string;
}
