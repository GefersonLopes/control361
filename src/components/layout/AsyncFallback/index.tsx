import React from "react";

import Spinner from "../../ui/Spinner";
import FallbackContainer from "../FallbackContainer";
import type { AsyncFallbackProps } from "./types";

const AsyncFallback: React.FC<AsyncFallbackProps> = ({
  isLoading,
  isError,
  errorContent = "Ocorreu um erro",
  loadingContent = <Spinner />,
  children,
  className,
}) => {
  if (isError) {
    return (
      <FallbackContainer className={className}>
        {errorContent}
      </FallbackContainer>
    );
  }

  if (isLoading) {
    return (
      <FallbackContainer className={className}>
        {loadingContent}
      </FallbackContainer>
    );
  }

  return <>{children}</>;
};

export default AsyncFallback;
