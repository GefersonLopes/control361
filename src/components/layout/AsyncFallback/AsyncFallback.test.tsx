import "@testing-library/jest-dom";

import { render, screen } from "@testing-library/react";

import AsyncFallback from "../AsyncFallback";

jest.mock("../../ui/Spinner", () => () => <div data-testid="spinner" />);
jest.mock(
  "../FallbackContainer",
  () =>
    ({
      className,
      children,
    }: {
      className: string;
      children: React.ReactNode;
    }) => (
      <div data-testid="fallback" className={className}>
        {children}
      </div>
    ),
);

describe("AsyncFallback component", () => {
  it("renders children when not loading or error", () => {
    render(
      <AsyncFallback isLoading={false} isError={false}>
        <span>Actual Content</span>
      </AsyncFallback>,
    );
    expect(screen.getByText("Actual Content")).toBeInTheDocument();
    expect(screen.queryByTestId("fallback")).toBeNull();
  });

  it("renders errorContent inside FallbackContainer when isError is true", () => {
    render(
      <AsyncFallback
        isLoading={false}
        isError={true}
        errorContent="Custom Error"
        className="error-class"
      >
        <span>Child</span>
      </AsyncFallback>,
    );
    const fallback = screen.getByTestId("fallback");
    expect(fallback).toHaveClass("error-class");
    expect(fallback).toHaveTextContent("Custom Error");
    expect(screen.queryByText("Child")).toBeNull();
  });

  it("uses default errorContent when isError and no errorContent provided", () => {
    render(
      <AsyncFallback isLoading={false} isError={true}>
        <span>Child</span>
      </AsyncFallback>,
    );
    const fallback = screen.getByTestId("fallback");
    expect(fallback).toHaveTextContent("Ocorreu um erro");
  });

  it("renders loadingContent inside FallbackContainer when isLoading is true and not error", () => {
    render(
      <AsyncFallback isLoading={true} isError={false} className="load-class">
        <span>Child</span>
      </AsyncFallback>,
    );
    const fallback = screen.getByTestId("fallback");
    expect(fallback).toHaveClass("load-class");
    expect(screen.getByTestId("spinner")).toBeInTheDocument();
    expect(screen.queryByText("Child")).toBeNull();
  });

  it("uses custom loadingContent when provided", () => {
    render(
      <AsyncFallback
        isLoading={true}
        isError={false}
        loadingContent={<span>Loading Now</span>}
      >
        <span>Child</span>
      </AsyncFallback>,
    );
    const fallback = screen.getByTestId("fallback");
    expect(fallback).toHaveTextContent("Loading Now");
    expect(screen.queryByText("Child")).toBeNull();
  });

  it("matches snapshot for error state", () => {
    const { asFragment } = render(
      <AsyncFallback isLoading={false} isError={true} errorContent="Err">
        <div>Child</div>
      </AsyncFallback>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for loading state", () => {
    const { asFragment } = render(
      <AsyncFallback isLoading={true} isError={false}>
        <div>Child</div>
      </AsyncFallback>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
