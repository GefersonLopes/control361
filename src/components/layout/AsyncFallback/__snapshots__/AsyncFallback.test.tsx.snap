// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncFallback component matches snapshot for error state 1`] = `
<DocumentFragment>
  <div
    data-testid="fallback"
  >
    Err
  </div>
</DocumentFragment>
`;

exports[`AsyncFallback component matches snapshot for loading state 1`] = `
<DocumentFragment>
  <div
    data-testid="fallback"
  >
    <div
      data-testid="spinner"
    />
  </div>
</DocumentFragment>
`;
